### PR TITLE
MM-39711 - Gekidou - Theme functionality

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
With this PR, we can set the app's theme.

Do not merge yet; waiting for PR #6312 .

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39711

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by larkox — APPROVED** [review]

File: app/screens/settings/display_theme/display_theme.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            return;
+        }
+        const pref: PreferenceType = {
+            category: Preferences.CATEGORY_THEME,
+            name: currentTeamId,
+            user_id: currentUserId,
+            value: JSON.stringify(Preferences.THEMES[selectedTheme]),
+        };
+        savePreference(serverUrl, [pref]);
+    }, [serverUrl]);
```
[the remark this answers is not in the document]
Comment (enahum) [reply app/screens/settings/display_theme/display_theme.tsx:53]: I would say, "Yes"

File: app/screens/settings/display_theme/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import DisplayTheme from './display_theme';
+
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
+    const currentTeamId = observeCurrentTeamId(database);
+    const currentUserId = observeCurrentUserId(database);
+    const themePreferences = queryPreferencesByCategoryAndName(database, Preferences.CATEGORY_THEME).observe();
+    const configAllowCustomThemes = observeConfigBooleanValue(database, 'AllowCustomThemes');
+
+    const customTheme = combineLatest([currentTeamId, themePreferences, configAllowCustomThemes]).pipe(
```
Comment (enahum) [inline app/screens/settings/display_theme/index.ts:28]: actually, I would simply this with the use of `useTheme` and inspecting if the type if `custom`, then for the ability to roll it back I would store that theme value in a local `customTheme` state

Comment (enahum) [reply app/screens/settings/display_theme/index.ts:28]: 🤯

File: app/screens/settings/display_theme/theme_thumbnail.tsx [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    borderColorMix = '#E0E1E3',
+    buttonBg = '#15B7B7',
+    centerChannelBg = 'white',
+    centerChannelColor = '#E0E1E3',
+    dndIndicator = '#D24B4E',
+    newMessageSeparator = '#1C58D9',
+    onlineIndicator = '#3DB887',
+    sidebarBg = '#174AB5',
+    sidebarText = '#86A1D9',
+    sidebarUnreadText = 'white',
```
Comment (enahum) [inline app/screens/settings/display_theme/theme_thumbnail.tsx:35]: what are this default values and why do we need them? was this just copied from v1? I would want to know exactly why we have them here.

File: app/screens/settings/display_theme/theme_thumbnail.tsx [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    borderColorMix: string;
+    buttonBg: string;
+    centerChannelBg: string;
+    centerChannelColor: string;
+    dndIndicator: string;
+    newMessageSeparator: string;
+    onlineIndicator: string;
+    sidebarBg: string;
+    sidebarText: string;
+    sidebarUnreadText: string;
```
Comment (enahum) [inline app/screens/settings/display_theme/theme_thumbnail.tsx:19]: I just not get why don't we pass the `theme` as the prop

File: app/screens/settings/display_theme/theme_tiles.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    label,
+    selected,
+    tileTheme,
+}: ThemeTileProps) => {
+    const style = getStyleSheet(activeTheme);
+    const {width: deviceWidth} = useWindowDimensions();
+
+    const layoutStyle = useMemo(() => {
+        const tilesPerLine = isLandscape || isTablet ? 4 : 2;
+        const fullWidth = isLandscape ? deviceWidth - 40 : deviceWidth;
```
Comment (enahum) [inline app/screens/settings/display_theme/theme_tiles.tsx:82]: isLandscape as only applied to tablets? I just don't want to see code brought from v1 without reasoning

File: assets/base/i18n/en.json [hunk 6]
```diff
@@ -555,7 +556,6 @@
   "notification_settings.auto_responder.default_message": "Hello, I am out of office and unable to respond to messages.",
   "notification_settings.auto_responder.enabled": "Enabled",
   "notification_settings.auto_responder.footer_message": "Set a custom message that will be automatically sent in response to Direct Messages. Mentions in Public and Private Channels will not trigger the automated reply. Enabling Automatic Replies sets your status to Out of Office and disables email and push notifications.",
-  "notification_settings.auto_responder.message_placeholder": "Message",
```
Comment (enahum) [inline assets/base/i18n/en.json:558]: @shazm see, not only my PR is removing this.. script bound

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/settings/display_theme/theme_tiles.tsx (hunk 5 above)
Comment (avinashlng1080) [reply app/screens/settings/display_theme/theme_tiles.tsx:82]: corrected

File: app/screens/settings/display_theme/theme_thumbnail.tsx (hunk 4 above)
Comment (avinashlng1080) [reply app/screens/settings/display_theme/theme_thumbnail.tsx:19]: corrected

File: app/screens/settings/display_theme/theme_thumbnail.tsx (hunk 3 above)
Comment (avinashlng1080) [reply app/screens/settings/display_theme/theme_thumbnail.tsx:35]: I didn't refactor it enough after porting... mindless porting.  Thanks for catching this :)

File: app/screens/settings/display_theme/index.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            const t = themePrefs.find((pref) => pref.name === teamId)?.value;
+            // eslint-disable-next-line max-nested-callbacks
+            const e = themePrefs.find((pref) => pref.name === '')?.value;
+
+            const themePref = t ? of$(t) : of$(e);
+            return themePref;
+        }),
+        switchMap((cth) => {
+            if (cth) {
+                const thx = safeParseJSON(cth) as unknown as Record<string, string>;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/settings/display_theme/index.ts]: I removed this part.

File: app/screens/settings/display_theme/display_theme.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/settings/display_theme/display_theme.tsx]: corrected

File: app/screens/index.tsx [hunk 8]
```diff
@@ -208,7 +214,6 @@ Navigation.setLazyComponentRegistrator((screenName) => {
             break;
         case Screens.USER_PROFILE:
             screen = withServerDatabase(require('@screens/user_profile').default);
-            break;
```
Comment (enahum) [inline app/screens/index.tsx:211]: why remove this break?

**Review by enahum — APPROVED** [review]

File: app/screens/index.tsx (hunk 8 above)
Comment (avinashlng1080) [reply app/screens/index.tsx:211]: corrected

File: app/screens/settings/display_theme/display_theme.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    const theme = useTheme();
+    const [customTheme, setCustomTheme] = useState<Theme|null>();
+
+    const styles = getStyleSheet(theme);
+
+    useEffect(() => {
+        if (theme.type === 'custom') {
+            setCustomTheme(theme);
+        }
+    }, [theme.type === 'custom']);
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/settings/display_theme/display_theme.tsx:46]: Yes. I'm going to remove the dependency.

[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/settings/display_theme/display_theme.tsx]: corrected

**Pull request merged**